python/pyspark/ml/image.py

```diff
@@ -0,0 +1,192 @@
+
+        .. versionadded:: 2.3.0
+        """
+
+        height = image.height
+        width = image.width
+        nChannels = image.nChannels
+        return np.ndarray(
+            shape=(height, width, nChannels),
+            dtype=np.uint8,
+            buffer=image.data,
+            strides=(width * nChannels, nChannels, 1))
+
+    def toImage(self, array, origin=""):
+        """
+        Converts a one-dimensional array to a two-dimensional image.
+
+        :param array array: The array to convert to image
+        :param str origin: Path to the image
+        :rtype object: Two dimensional image
+
+        .. versionadded:: 2.3.0
+        """
+
+        if array.ndim != 3:
+            raise ValueError("Invalid array shape")
+        height, width, nChannels = array.shape
+        ocvTypes = ImageSchema.ocvTypes
+        if nChannels == 1:
+            mode = ocvTypes["CV_8UC1"]
+        elif nChannels == 3:
+            mode = ocvTypes["CV_8UC3"]
+        elif nChannels == 4:
+            mode = ocvTypes["CV_8UC4"]
+        else:
+            raise ValueError("Invalid number of channels")
+        data = bytearray(array.astype(dtype=np.uint8).ravel())
+        # Creating new Row with _create_row(), because Row(name = value, ... )
+        # orders fields by name, which conflicts with expected schema order
+        # when the new DataFrame is created by UDF
+        return _create_row(self.imageFields,
+                           [origin, height, width, nChannels, mode, data])
+
+    def readImages(self, path, recursive=False, numPartitions=0,
+                   dropImageFailures=False, sampleRatio=1.0):
+        """
+        Reads the directory of images from the local or remote source.
+
+        :param str path: Path to the image directory
+        :param SparkSession spark: The current spark session
+        :param bool recursive: Recursive search flag
+        :param int numPartitions: Number of DataFrame partitions
+        :param bool dropImageFailures: Drop the files that are not valid images
+        :param float sampleRatio: Fraction of the images loaded
+        :rtype DataFrame: DataFrame with a single column of "images",
+               see ImageSchema for details
+
+        >>> df = ImageSchema.readImages('python/test_support/image/kittens', recursive=True)
+        >>> df.count()
+        4
+
+        .. versionadded:: 2.3.0
+        """
+
+        ctx = SparkContext._active_spark_context
+        spark = SparkSession(ctx)
+        image_schema = ctx._jvm.org.apache.spark.ml.image.ImageSchema
+        jsession = spark._jsparkSession
+        jresult = image_schema.readImages(path, jsession, recursive, numPartitions,
+                                          dropImageFailures, float(sampleRatio))
+        return DataFrame(jresult, spark._wrapped)
+
+
+ImageSchema = _ImageSchema()
+
+
+# Monkey patch to disallow instantization of this class.
+def _disallow_instance(_):
```

Comment:
If this monkey patch worries anyone, I think we could just remove it because the underscore in `_ImageSchema` already indicates it's private. It's rather a hacky trick and I only added this because the doc shows this class.

Reply:
+0 to removing the monkey patch (but not a strong feeling)

Reply:
sorry if I misunderstand here, but +0 means we should keep the monkey patch (as opposed to +1 for removing it) - so I won't make any change here for now, please let me know if otherwise